File: apps/web/src/lib/analytics/highlights.ts
import type { Entry } from '../../api/types'
import type { DayHighlight, CashOutDay } from './types'

type DayData = {
  date: string
  income: number
  expense: number
  balance: number
}

function aggregateByDay(entries: Entry[]): Map<string, DayData> {
  const byDate = new Map<string, DayData>()

  for (const e of entries) {
    const date = e.TransactionDate.slice(0, 10)
    let day = byDate.get(date)
    if (!day) {
      day = { date, income: 0, expense: 0, balance: 0 }
      byDate.set(date, day)
    }
    if (e.Type === 'income' && e.Category === 'venda_balcao') {
      day.income += e.Amount
    } else {
      day.expense += e.Amount
    }
    day.balance = day.income - day.expense
  }

  return byDate
}

export function getHighlights(entries: Entry[]): {
  bestIncome: DayHighlight
  worstIncome: DayHighlight
  bestBalance: DayHighlight
  worstBalance: DayHighlight
} {
  const byDate = aggregateByDay(entries)
  const days = Array.from(byDate.values())

  const empty: DayHighlight = { date: '—', label: 'Sem dados', amount: 0 }

  if (days.length === 0) {
    return {
      bestIncome: empty,
      worstIncome: empty,
      bestBalance: empty,
      worstBalance: empty,
    }
  }

  const bestIncome = days.reduce((best, d) =>
    d.income > best.income ? d : best,
  )
  const worstIncome = days.reduce((worst, d) =>
    d.income < worst.income ? d : worst,
  )
  const bestBalance = days.reduce((best, d) =>
    d.balance > best.balance ? d : best,
  )
  const worstBalance = days.reduce((worst, d) =>
    d.balance < worst.balance ? d : worst,
  )

  return {
    bestIncome: toHighlight(bestIncome, 'income'),
    worstIncome: toHighlight(worstIncome, 'income'),
    bestBalance: toHighlight(bestBalance, 'balance'),
    worstBalance: toHighlight(worstBalance, 'balance'),
  }
}

function toHighlight(
  day: DayData,
  type: 'income' | 'balance',
): DayHighlight {
  const amount = type === 'income' ? day.income : day.balance
  const date = new Date(day.date + 'T12:00:00')
  const label = date.toLocaleDateString('pt-BR', {
    day: '2-digit',
    month: 'short',
  })
  return { date: day.date, label, amount }
}

export function getCashOutDays(entries: Entry[]): CashOutDay[] {
  const byDate = new Map<string, { total: number; items: Map<string, { amount: number; count: number }> }>()

  for (const e of entries) {
    if (e.Type !== 'expense') continue
    const date = e.TransactionDate.slice(0, 10)
    let bucket = byDate.get(date)
    if (!bucket) {
      bucket = { total: 0, items: new Map() }
      byDate.set(date, bucket)
    }
    bucket.total += e.Amount
    const existing = bucket.items.get(e.Category)
    if (existing) {
      existing.amount += e.Amount
      existing.count += 1
    } else {
      bucket.items.set(e.Category, { amount: e.Amount, count: 1 })
    }
  }

  return Array.from(byDate.entries())
    .map(([date, data]) => ({
      date,
      total: data.total,
      items: Array.from(data.items.entries())
        .map(([category, v]) => ({ category, amount: v.amount, count: v.count }))
        .sort((a, b) => b.amount - a.amount),
    }))
    .sort((a, b) => b.total - a.total)
    .slice(0, 5)
}

File: packages/finance/analytics/highlights.go
package analytics

import (
	"sort"

	"github.com/emerson/emerbot/packages/domain"
)

// maxCashOutDays caps how many heavy-spending days the analysis reports.
const maxCashOutDays = 5

type dayTotals struct {
	date string
	// revenue is faturamento only (see domain.IsRevenue) — what BestIncome and
	// WorstIncome are measured against, so a loan or aporte can't manufacture
	// a "best sales day".
	revenue int64
	// cashIn is every inflow, sale or not — what balance is measured against,
	// since a real cash movement must not be dropped just because it wasn't a
	// sale.
	cashIn  int64
	expense int64
	balance int64
}

// aggregateByDay folds entries into per-day totals, ordered oldest-first.
//
// It takes two slices because the two figures are measured on two different
// bases: revenueEntries on the transaction basis (a sale belongs to the day it
// was made) and entries on the effective-date basis (everything else). Passing
// the same slice twice would silently put an unpaid crediário sale on the wrong
// day.
func aggregateByDay(entries, revenueEntries []domain.FinancialEntry) []dayTotals {
	byDate := map[string]*dayTotals{}
	day := func(date string) *dayTotals {
		d, ok := byDate[date]
		if !ok {
			d = &dayTotals{date: date}
			byDate[date] = d
		}
		return d
	}
	for _, e := range revenueEntries {
		if domain.IsRevenue(e) {
			day(e.TransactionDate.String()).revenue += e.Amount
		}
	}
	for _, e := range entries {
		d := day(e.TransactionDate.String())
		if e.Type == domain.EntryTypeIncome {
			d.cashIn += e.Amount
		} else {
			d.expense += e.Amount
		}
		d.balance = d.cashIn - d.expense
	}

	days := make([]dayTotals, 0, len(byDate))
	for _, d := range byDate {
		days = append(days, *d)
	}
	// Sorted so ties resolve to the earliest day rather than to whatever order
	// the map happened to hand back.
	sort.Slice(days, func(i, j int) bool { return days[i].date < days[j].date })
	return days
}

// buildHighlights picks the best and worst day of the month by income and by
// balance. With no entries at all every slot is the same "Sem dados"
// placeholder, so the dashboard renders empty cards instead of a zero-value
// date.
func buildHighlights(entries, revenueEntries []domain.FinancialEntry) Highlights {
	days := aggregateByDay(entries, revenueEntries)
	if len(days) == 0 {
		empty := DayHighlight{Date: NoDataDate, Label: "Sem dados", Amount: 0}
		return Highlights{BestIncome: empty, WorstIncome: empty, BestBalance: empty, WorstBalance: empty}
	}

	bestIncome, worstIncome := days[0], days[0]
	bestBalance, worstBalance := days[0], days[0]
	for _, d := range days[1:] {
		if d.revenue > bestIncome.revenue {
			bestIncome = d
		}
		if d.revenue < worstIncome.revenue {
			worstIncome = d
		}
		if d.balance > bestBalance.balance {
			bestBalance = d
		}
		if d.balance < worstBalance.balance {
			worstBalance = d
		}
	}

	return Highlights{
		BestIncome:   toHighlight(bestIncome, bestIncome.revenue),
		WorstIncome:  toHighlight(worstIncome, worstIncome.revenue),
		BestBalance:  toHighlight(bestBalance, bestBalance.balance),
		WorstBalance: toHighlight(worstBalance, worstBalance.balance),
	}
}

func toHighlight(day dayTotals, amount int64) DayHighlight {
	label := day.date
	if d, err := domain.ParseCalendarDate(day.date); err == nil {
		label = dayMonthLabel(d.Time())
	}
	return DayHighlight{Date: day.date, Label: label, Amount: amount}
}

// buildCashOutDays returns the days with the heaviest outgoings, broken down
// by category, so "where did the money go?" has an answer without opening the
// entries table.
func buildCashOutDays(entries []domain.FinancialEntry) []CashOutDay {
	type acc struct {
		total int64
		items map[string]*CashOutItem
	}
	byDate := map[string]*acc{}

	for _, e := range entries {
		if e.Type != domain.EntryTypeExpense {
			continue
		}
		date := e.TransactionDate.String()
		bucket, ok := byDate[date]
		if !ok {
			bucket = &acc{items: map[string]*CashOutItem{}}
			byDate[date] = bucket
		}
		bucket.total += e.Amount
		item, ok := bucket.items[e.Category]
		if !ok {
			item = &CashOutItem{Category: e.Category}
			bucket.items[e.Category] = item
		}
		item.Amount += e.Amount
		item.Count++
	}

	days := make([]CashOutDay, 0, len(byDate))
	for date, bucket := range byDate {
		items := make([]CashOutItem, 0, len(bucket.items))
		for _, item := range bucket.items {
			items = append(items, *item)
		}
		sort.Slice(items, func(i, j int) bool {
			if items[i].Amount != items[j].Amount {
				return items[i].Amount > items[j].Amount
			}
			return items[i].Category < items[j].Category
		})
		days = append(days, CashOutDay{Date: date, Total: bucket.total, Items: items})
	}

	// Heaviest first; the date breaks ties so the top-N cut is stable rather
	// than dependent on map ordering.
	sort.Slice(days, func(i, j int) bool {
		if days[i].Total != days[j].Total {
			return days[i].Total > days[j].Total
		}
		return days[i].Date < days[j].Date
	})
	if len(days) > maxCashOutDays {
		days = days[:maxCashOutDays]
	}
	return days
}
